<a href="https://colab.research.google.com/github/vperls000/csc44800/blob/main/Viviana_Perls_Assignment4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/vperls000/csc44800/blob/main/Viviana_Perls_Assignment4.ipynb)

In [ ]:
# Imports and pip installations (if needed)
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn import svm
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier
import pandas as pd

# Part 1: Load the dataset

In [ ]:
# Load the dataset (load remotely, not locally)
csv_url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'
column_names = ['sepal length', 'sepal width', 'petal length', 'petal width', 'species']
iris = pd.read_csv(csv_url,names=column_names)

# Output the first 15 rows of the data
iris.head(15)

sepal length  sepal width  petal length  petal width      species
0            5.1          3.5           1.4          0.2  Iris-setosa
1            4.9          3.0           1.4          0.2  Iris-setosa
2            4.7          3.2           1.3          0.2  Iris-setosa
3            4.6          3.1           1.5          0.2  Iris-setosa
4            5.0          3.6           1.4          0.2  Iris-setosa
5            5.4          3.9           1.7          0.4  Iris-setosa
6            4.6          3.4           1.4          0.3  Iris-setosa
7            5.0          3.4           1.5          0.2  Iris-setosa
8            4.4          2.9           1.4          0.2  Iris-setosa
9            4.9          3.1           1.5          0.1  Iris-setosa
10           5.4          3.7           1.5          0.2  Iris-setosa
11           4.8          3.4           1.6          0.2  Iris-setosa
12           4.8          3.0           1.4          0.1  Iris-setosa
13           4.3          3.0           1.1          0.1  Iris-setosa
14           5.8          4.0           1.2          0.2  Iris-setosa

In [ ]:
# Display a summary of the table information (number of datapoints, etc.)
print(iris.describe())

       sepal length  sepal width  petal length  petal width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.054000      3.758667     1.198667
std        0.828066     0.433594      1.764420     0.763161
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000


## About the dataset
The Iris Dataset is a commonly used toy dataset which contains 150 samples of irises of the species Setosa, Versicolor, and Virginica- where each species has 50 samples within the dataset.  The species will be classified into groups {0,1,2}, where 0 = Setosa, 1 = Versicolor, and 2 = Virginica.  The features of these irises are sepal length, sepal width, petal length, and petal width.

# Part 2: Split the dataset into train and test

In [ ]:
# Take the dataset and split it into our features (X) and label (y)
X = iris[['sepal length', 'sepal width', 'petal length', 'petal width']]
Y = iris['species']

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=.1, train_size=.9, random_state=0, shuffle=True, stratify=None)

In [ ]:
# Here I am defining the sample datapoint that I will be using to predict the
# probabilities for each possible class in each model.
X_sample = [[7.900000,4.400000,6.900000,2.500000]]
X_pred = pd.DataFrame(X_sample, columns=X_test.columns)

# Part 3: Logistic Regression

In [ ]:
# i. Use sklearn to train a LogisticRegression model on the training set
reg = LogisticRegression(random_state=0).fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
y_pred_reg = reg.predict_proba(X_pred)
print('Setosa probability: ', y_pred_reg[0][0])
print('Versicolor probability: ', y_pred_reg[0][1])
print('Virginica probability: ', y_pred_reg[0][2])

# iii. Report on the score for Logistic regression model, what does the score measure?
print('Score: ', reg.score(X_test,y_test))

# iv. Extract the coefficents and intercepts for the boundary line(s)
print('Coefficients: ', reg.coef_)
print('Intercepts: ', reg.intercept_)

Setosa probability:  3.7740018393379437e-08
Versicolor probability:  0.001185918144024011
Virginica probability:  0.9988140441159576
Score:  1.0
Coefficients:  [[-0.41035194  0.92087293 -2.43647188 -1.05570355]
 [ 0.40249086 -0.32260361 -0.11298067 -0.88152875]
 [ 0.00786108 -0.59826933  2.54945255  1.9372323 ]]
Intercepts:  [  9.49196592   2.42958512 -11.92155104]


The score of the Logistic Regression model is the mean accuracy of the model given the test data.  In this instance, the score is 1, meaning that (on average) the model can classify data from X_test with 100% accuracy.

# Part 4: Support Vector Machine

In [ ]:
# i. Use sklearn to train a Support Vector Classifier on the training set
svc = svm.SVC(probability=True).fit(X_train,y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
y_pred_svc = svc.predict_proba(X_pred)
print('Setosa probability: ', y_pred_svc[0][0])
print('Versicolor probability: ', y_pred_svc[0][1])
print('Virginica probability: ', y_pred_svc[0][2])

# iii. Report on the score for the SVM, what does the score measure?
print('Score: ', svc.score(X_test,y_test))

Setosa probability:  0.028066488976037213
Versicolor probability:  0.007891656689803603
Virginica probability:  0.9640418543341593
Score:  1.0


The score measures the same mean accuracy as in the Logistic Regression model, as it does in all of the models in this assignment.  This score is also 1, meaning that the accuracy is 100%.  However, looking at the prediction probabilities for each possible class, the same sample has a probability of 96% for the Virginica species, and 99% of the same in the Logistic Regression model.  Therefore, the two have the same mean accuracy, but perhaps the Logistic Regression model would be more accurate given a larger set of test data.

# Part 5: Neural Network

In [ ]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
net = MLPClassifier(random_state=0,max_iter=750).fit(X_train,y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
y_pred_net = net.predict_proba(X_pred)
print('Setosa probability: ', y_pred_net[0][0])
print('Versicolor probability: ', y_pred_net[0][1])
print('Virginica probability: ', y_pred_net[0][2])

# iii. Report on the score for the Neural Network, what does the score measure?
print('Score: ',net.score(X_test,y_test))

Setosa probability:  5.006597396348999e-07
Versicolor probability:  0.0036177365516225865
Virginica probability:  0.9963817627886379
Score:  1.0


The score is again 1 in this Neutal Net model, which again means 100% accuracy.  For the individual sample, the predicted probability of the Virginica species is 99%, indicating more similarity to the Logistic Regression model than to the Support Vector Machine model.

In [ ]:
# Experimenting with max_iter value

net = MLPClassifier(random_state=0,max_iter=300).fit(X_train,y_train)
print('Score when max_iter=300: ',net.score(X_test,y_test))

net = MLPClassifier(random_state=0,max_iter=200).fit(X_train,y_train)
print('Score when max_iter=200: ',net.score(X_test,y_test))

net = MLPClassifier(random_state=0,max_iter=100).fit(X_train,y_train)
print('Score when max_iter=100: ',net.score(X_test,y_test))

net = MLPClassifier(random_state=0,max_iter=110).fit(X_train,y_train)
print('Score when max_iter=110: ',net.score(X_test,y_test))

net = MLPClassifier(random_state=0,max_iter=115).fit(X_train,y_train)
print('Score when max_iter=115: ',net.score(X_test,y_test))

net = MLPClassifier(random_state=0,max_iter=120).fit(X_train,y_train)
print('Score when max_iter=120: ',net.score(X_test,y_test))

/usr/local/lib/python3.7/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:696: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (300) reached and the optimization hasn't converged yet.
  ConvergenceWarning,


Score when max_iter=300:  1.0


/usr/local/lib/python3.7/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:696: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:696: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (100) reached and the optimization hasn't converged yet.
  ConvergenceWarning,


Score when max_iter=200:  1.0
Score when max_iter=100:  0.8


/usr/local/lib/python3.7/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:696: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (110) reached and the optimization hasn't converged yet.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:696: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (115) reached and the optimization hasn't converged yet.
  ConvergenceWarning,


Score when max_iter=110:  0.8666666666666667
Score when max_iter=115:  0.9333333333333333
Score when max_iter=120:  1.0


/usr/local/lib/python3.7/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:696: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (120) reached and the optimization hasn't converged yet.
  ConvergenceWarning,


In the original Neural Net model, I set max_iter to 750 to avoid a convergence warning.  In the two models above, I lowered the number of max iterations to see at about what point the score dips below 1.  It seems that past about 115 iterations is when the model becomes 100% accurate.

In [ ]:
# Experimenting with hidden_layer_sizes
# max_iter = 750

net = MLPClassifier(random_state=0,max_iter=750,hidden_layer_sizes=1).fit(X_train,y_train)
print('Score when hidden_layer_sizes=1: ', net.score(X_test,y_test))

net = MLPClassifier(random_state=0,max_iter=750,hidden_layer_sizes=5).fit(X_train,y_train)
print('\nScore when hidden_layer_sizes=5: ', net.score(X_test,y_test))

net = MLPClassifier(random_state=0,max_iter=750,hidden_layer_sizes=10).fit(X_train,y_train)
print('\nScore when hidden_layer_sizes=10: ', net.score(X_test,y_test))

/usr/local/lib/python3.7/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:696: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (750) reached and the optimization hasn't converged yet.
  ConvergenceWarning,


Score when hidden_layer_sizes=1:  0.4666666666666667


/usr/local/lib/python3.7/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:696: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (750) reached and the optimization hasn't converged yet.
  ConvergenceWarning,


Score when hidden_layer_sizes=5:  0.8666666666666667
Score when hidden_layer_sizes=10:  1.0


/usr/local/lib/python3.7/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:696: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (750) reached and the optimization hasn't converged yet.
  ConvergenceWarning,


I repeated the process by changing the number of hidden layers to see when the score dips below 1, making sure to keep max_iter at 750 for each model.  From these tests, I noticed that the range of hidden layer sizes was much smaller here than the range of max iterations.

In [ ]:
# Experimenting with both max_iter and hidden_layer_size

net= MLPClassifier(random_state=0,max_iter=1000,hidden_layer_sizes=1).fit(X_train,y_train)
print('Score when max_iter=1000, hidden_layer_sizes=1: ', net.score(X_test,y_test))

net= MLPClassifier(random_state=0,max_iter=2000,hidden_layer_sizes=1).fit(X_train,y_train)
print('Score when max_iter=2000, hidden_layer_sizes=1: ', net.score(X_test,y_test))

net= MLPClassifier(random_state=0,max_iter=3000,hidden_layer_sizes=1).fit(X_train,y_train)
print('Score when max_iter=3000, hidden_layer_sizes=1: ', net.score(X_test,y_test))

net= MLPClassifier(random_state=0,max_iter=1000,hidden_layer_sizes=5).fit(X_train,y_train)
print('Score when max_iter=1000, hidden_layer_sizes=5: ', net.score(X_test,y_test))

/usr/local/lib/python3.7/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:696: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  ConvergenceWarning,


Score when max_iter=1000, hidden_layer_sizes=1:  0.4666666666666667


/usr/local/lib/python3.7/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:696: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (2000) reached and the optimization hasn't converged yet.
  ConvergenceWarning,


Score when max_iter=2000, hidden_layer_sizes=1:  0.6
Score when max_iter=3000, hidden_layer_sizes=1:  0.6
Score when max_iter=1000, hidden_layer_sizes=5:  1.0


/usr/local/lib/python3.7/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:696: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  ConvergenceWarning,


In the tests where I only changed the hidden layer sizes, I noticed that I got convergence warnings where I hadn't in the original model, so I decided to try increasing the number of max iterations until the score changes or until the convergence warning went away.  From the tests above, it's clear that the less hidden layers there are, the more iterations may be necessary to avoid a convergence warning or to get a good score.

# Part 6: K-Nearest Neighbors

In [ ]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier

knn = KNeighborsClassifier().fit(X_train,y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
y_pred_knn = knn.predict_proba(X_pred)
print('Setosa probability: ', y_pred_knn[0][0])
print('Versicolor probability: ', y_pred_knn[0][1])
print('Virginica probability: ', y_pred_knn[0][2])

# iii. Report on the score for kNN, what does the score measure?
print('\nScore: ', knn.score(X_test,y_test))

Setosa probability:  0.0
Versicolor probability:  0.0
Virginica probability:  1.0

Score:  1.0


The score for this K-Nearest Neighbors model was 1, but even more interestingly the prediction probabilities for the individual sample gave the Virginica species 100% probability.  I would assume that the sample I used (which was also the maximum from the summary table) only had neighbors that were of the Virginica species.

# Part 7: Conclusions and takeaways

Most of the models had a score of 1, which I expect is because the ratio of training to testing data is so high.  Based on the predicted probabilities of the individual sample, I would say that Logistic Regression and Support Machine worked best.  I am tempted to say the same for K-Nearest Neighbors, but I am not sure that this would be true if the sample I used was the mean from the summary.  I think KNN performed so well with my tests because my tests included very little data and therefore did not show what would happen if a datapoint was sort of in the middle of the distribution.  I was surprised to see how accurate every model was, but again I think this had to do with the small amount of test data in comparison to the training data.